chainstate/tx-verifier/src/transaction_verifier/input_output_policy/constraints_accumulator.rs

```diff
@@ -13,705 +13,584 @@
-            Some(TxOutput::CreateStakePool(
-                pool_id,
-                Box::new(stake_pool_data),
-            )),
-            Some(TxOutput::Transfer(
-                OutputValue::Coin(less_than_staked_amount),
-                Destination::AnyoneCanSpend,
-            )),
-        ];
-
-        // it's an error if output includes staked coins
-        let outputs = vec![TxOutput::Transfer(
-            OutputValue::Coin(Amount::from_atoms(staked_atoms)),
-            Destination::AnyoneCanSpend,
-        )];
-
-        {
-            let mut constraints_accumulator = ConstrainedValueAccumulator::new();
-            constraints_accumulator
-                .process_inputs(
-                    &chain_config,
-                    BlockHeight::new(1),
-                    pledge_getter,
-                    &inputs,
-                    &inputs_utxos,
-                )
-                .unwrap();
-
-            let result =
-                constraints_accumulator.process_outputs(&chain_config, &outputs).unwrap_err();
-            assert_eq!(
-                result,
-                IOPolicyError::AttemptToPrintMoneyOrViolateTimelockConstraints
-            );
+    pub fn subtract(&mut self, other: ConstrainedValueAccumulator) -> Result<(), IOPolicyError> {
+        for (key, value) in other.unconstrained_value {
+            decrease_or(
+                &mut self.unconstrained_value,
+                key,
+                value,
+                IOPolicyError::AttemptToPrintMoneyOrViolateTimelockConstraints(key),
+            )?;
         }
 
-        // it's not an error if output does not include staked coins
-        let outputs = vec![TxOutput::Transfer(
-            OutputValue::Coin(less_than_staked_amount),
-            Destination::AnyoneCanSpend,
-        )];
-
-        {
-            let mut constraints_accumulator = ConstrainedValueAccumulator::new();
-            constraints_accumulator
-                .process_inputs(
-                    &chain_config,
-                    BlockHeight::new(1),
-                    pledge_getter,
-                    &inputs,
-                    &inputs_utxos,
-                )
-                .unwrap();
-
-            constraints_accumulator.process_outputs(&chain_config, &outputs).unwrap();
+        for (timelock, locked_coins) in other.timelock_constrained {
+            let mut constraint_range_iter = self
+                .timelock_constrained
+                .range_mut((
+                    std::ops::Bound::Unbounded,
+                    std::ops::Bound::Included(timelock),
+                ))
+                .rev()
+                .peekable();
+
+            // iterate over the range until current output coins are completely used
+            // or all suitable constraints are satisfied
+            let mut output_coins = locked_coins;
+            while output_coins > Amount::ZERO {
+                match constraint_range_iter.peek_mut() {
```

Comment:
I am not quite convinced by the algorithm to subtract time-locked coins. We should check (probably using a randomized test) that the usual properties we would expect from an additive group hold, for exmaple:
* (A - B) + (C - D) = (A + C) - (B + D)
* A - (B + C) = A - B - C
* A - B = A + (0 - B)

I suspect they do not. If that's indeed the case, we should either:
* make the subtraction return a different type (maybe the resulting fees) so it's not being used in subsequent subsequent operations, or
* find a representation and algorithm that follow some sort of sensible set of algebraic identities

Reply:
Good point, thanks. I added separate type and renamed `subtract` method to `satisfy_with` to avoid confusion.